src/librustc/infer/error_reporting/named_anon_conflict.rs

```diff
@@ -0,0 +1,115 @@
+    // by that of the named region.
+    pub fn find_arg_with_anonymous_region(&self,
+                                          anon_region: Region<'tcx>,
+                                          named_region: Region<'tcx>)
+                                          -> Option<(&hir::Arg, ty::Ty<'tcx>)> {
+
+        match *anon_region {
+            ty::ReFree(ref free_region) => {
+
+                let id = free_region.scope;
+                let node_id = self.tcx.hir.as_local_node_id(id).unwrap();
+                let body_id = self.tcx.hir.maybe_body_owned_by(node_id).unwrap();
+
+                let body = self.tcx.hir.body(body_id);
+                body.arguments
+                    .iter()
+                    .filter_map(|arg| if let Some(tables) = self.in_progress_tables {
+                                    let ty = tables.borrow().node_id_to_type(arg.id);
+                                    let mut found_anon_region = false;
+                                    let new_arg_ty = self.tcx
+                                        .fold_regions(&ty,
+                                                      &mut false,
+                                                      |r, _| if *r == *anon_region {
+                                                          found_anon_region = true;
+                                                          named_region
+                                                      } else {
+                                                          r
+                                                      });
+                                    if found_anon_region {
+                                        return Some((arg, new_arg_ty));
+                                    } else {
+                                        None
+                                    }
+                                } else {
+                                    None
+                                })
+                    .next()
+            }
+            _ => None,
+        }
+
+    }
+
+    // This method generates the error message for the case when
+    // the function arguments consist of a named region and an anonymous
+    // region and corresponds to `ConcreteFailure(..)`
+    pub fn report_named_anon_conflict(&self, error: &RegionResolutionError<'tcx>) -> bool {
+
+        let (span, sub, sup) = match *error {
+            ConcreteFailure(ref origin, sub, sup) => (origin.span(), sub, sup),
+            _ => return false, // inapplicable
+        };
+
+        let (named, (var, new_ty)) =
+            if self.is_named_region(sub) && self.is_anonymous_region(sup) {
+                (sub, self.find_arg_with_anonymous_region(sup, sub).unwrap())
+            } else if self.is_named_region(sup) && self.is_anonymous_region(sub) {
+                (sup, self.find_arg_with_anonymous_region(sub, sup).unwrap())
+            } else {
+                return false; // inapplicable
+            };
+
```

Comment:
So, per our conversation earlier, I think we want to watch out for the case where the anonymous region appeared in the function **return type** as well and not try to handle that (that would be #42703 now). To do this check is a bit of a pain though. I think we need to do something like this:

- extract the `free_region.scope` from the anonymous region (just as we do in `find_arg_with_anonymous_region()`); let's call that `scope_def_id,`
- lookup the type `ty` associated with `scope_def_id,` using `tcx.type_of(scope_def_id)`
    - since we are only matching functions, this will be a `TyFnDef` (we can just return for other types)
    - that is the type of a named function like `fn foo()`
    - if we do `match ty.sty { ty::TyFnDef(_, _, sig) => ... }` we can extract the function signature `sig` (it is a `PolyFnSig`)
    - we can extract the return type `ret_ty` by doing this incantation (I can explain more on gitter): `tcx.liberate_late_bound_regions(scope_def_id, sig.output())`
- then we can search `ret_ty` for the anonymous region as we did before (something like `ret_ty.walk().flat_map(|t| t.regions()).any(|r| r == anon_region)`